pinot-segment-spi/src/main/java/org/apache/pinot/segment/spi/index/creator/JsonIndexers.java

```diff
@@ -0,0 +1,54 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.pinot.segment.spi.index.creator;
+
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.function.Function;
+
+
+/**
+ * Registration point to allow overriding of JsonIndexing
+ */
+public final class JsonIndexers {
+
+  private JsonIndexers() {
+  }
+
+  private static final DefaultJsonIndexer DEFAULT_JSON_INDEXER = new DefaultJsonIndexer();
+  private static final Function<JsonIndexer, JsonIndexer> DEFAULT = Function.identity();
+  private static final AtomicReference<Function<JsonIndexer, JsonIndexer>> REGISTERED = new AtomicReference<>(DEFAULT);
+
+  /**
+   * Register a decorator which gets wraps the default JSON indexer so that it can intercept requests to
```

Comment:
Why do we need to intercept requests to index JSON column? Just registering the indexer should be sufficient, right?

Reply:
Currently its assumed that whatever is in the JSON column is actually JSON, but the indexing technique can be made generic (so that you could have a forward index containing, say, Protobuf records stored in bytes, and then a JSON index so long as you have a plugin which can construct the appropriate paths) with the layer of indirection provided in this PR.

Reply:
We don't need to assume that. Whatever is in the `JSON` datatype column has to be JSON. The reason we want to do this is to make subsequent query processing and indexing simpler. If we introduce Protobuf support later on, I really hope that we would do so by introducing a new datatype as well. JSON specification is very specific, so we should not overload the use of term JSON to mean things that JSON specification isn't.

The plugin concept, in my view, would be useful to have different implementations of JSON index. For example a tree based JSON index or a Hashtable based JSON index which could help in optimizing particular types of JSON queries. However, we should keep these a JSON index, separate from ComplexType index, or a Protobuf index.

Reply:
I don't think that the JSON data type constrains the physical representation of the JSON. With the right plugins, I should be able to store destructured JSON in a JSON column, but if I have destructured JSON, I might not want to rematerialise it as JSON just so it can be indexed.

Reply:
It seems like you have more things in mind than what is coming through across this PR and your last PR on JSON. It almost seems like we should have a design document here so that the path you are perusing is clear.

My original question still remains unanswered: Why do we need to intercept requests to index JSON column? Just registering the indexer should be sufficient, right?

> With the right plugins, I should be able to store destructured JSON in a JSON column

With the right "storage" plugin, you could store destructured JSON into a JSON column or store destructed JSON into multiple columns, etc. An "indexing" plugin may then be associated with a corresponding storage plugin if needed. (and still not sure where protobuf comes into this).

I really think we need to take the time to create a design document here, as to what you are trying to do and what the end goal is, before we move forward with further changes.

Reply:
> My original question still remains unanswered: Why do we need to intercept requests to index JSON column? Just registering the indexer should be sufficient, right?

I tried to answer this by way of example. Currently it is assumed that `ForwardIndexReader#getString` returns a JSON document for a JSON column, which constrains the physical representation. This PR provides a pluggable mediation mechanism between the `ForwardIndexReader` and the `JsonIndexCreator`.

Reply:
I am not seeing the purpose behind `Function<JsonIndexer, JsonIndexer>`, I think that can be avoided. Happy to discuss offline if needed.